cub/device/dispatch/dispatch_batch_memcpy.cuh

```diff
@@ -0,0 +1,631 @@
+      MEM_BLEV_BLOCK_SCAN_STATE,
+      // Total number of distinct memory allocations in the temporary storage memory BLOB
+      MEM_NUM_ALLOCATIONS
+    };
+
+    // Number of threads per block for initializing the grid states
+    constexpr uint32_t INIT_KERNEL_THREADS = 128U;
+
+    // The number of buffers that get processed per thread block
+    constexpr uint32_t TILE_SIZE = ActivePolicyT::AgentSmallBufferPolicyT::BLOCK_THREADS *
+                                   ActivePolicyT::AgentSmallBufferPolicyT::BUFFERS_PER_THREAD;
+
+    // The number of thread blocks (or tiles) required to process all of the given buffers
+    BlockOffsetT num_tiles = DivideAndRoundUp(num_buffers, TILE_SIZE);
+
+    // Temporary storage allocation requirements
+    void *allocations[MEM_NUM_ALLOCATIONS]       = {};
+    size_t allocation_sizes[MEM_NUM_ALLOCATIONS] = {};
+
+    using BlevBufferSrcsOutItT        = void **;
+    using BlevBufferDstsOutItT        = void **;
+    using BlevBufferSizesOutItT       = BufferSizeT *;
+    using BlevBufferTileOffsetsOutItT = BlockOffsetT *;
+
+    allocation_sizes[MEM_BLEV_BUFFER_SRCS]   = num_buffers * sizeof(void *);
+    allocation_sizes[MEM_BLEV_BUFFER_DSTS]   = num_buffers * sizeof(void *);
+    allocation_sizes[MEM_BLEV_BUFFER_SIZES]  = num_buffers * sizeof(BufferSizeT);
+    allocation_sizes[MEM_BLEV_BUFFER_TBLOCK] = num_buffers * sizeof(BlockOffsetT);
+    CubDebug(
+      error = BLevBufferOffsetTileState::AllocationSize(num_tiles, allocation_sizes[MEM_BLEV_BUFFER_SCAN_STATE]));
+    if (error)
+    {
+      return error;
+    }
+    CubDebug(error = BLevBlockOffsetTileState::AllocationSize(num_tiles, allocation_sizes[MEM_BLEV_BLOCK_SCAN_STATE]));
+    if (error)
+    {
+      return error;
+    }
+
+    // Compute allocation pointers into the single storage BLOB (or compute the necessary size of the blob)
+    AliasTemporaries<MEM_NUM_ALLOCATIONS>(d_temp_storage, temp_storage_bytes, allocations, allocation_sizes);
+
+    // Just return if no temporary storage is provided
+    if (d_temp_storage == nullptr)
+    {
+      return error;
+    }
+
+    // Return if empty problem
+    if (num_buffers == 0)
+    {
+      return error;
+    }
+
+    // Alias into temporary storage allocation
+    BlevBufferSrcsOutItT d_blev_src_buffers = reinterpret_cast<BlevBufferSrcsOutItT>(allocations[MEM_BLEV_BUFFER_SRCS]);
+    BlevBufferDstsOutItT d_blev_dst_buffers = reinterpret_cast<BlevBufferDstsOutItT>(allocations[MEM_BLEV_BUFFER_DSTS]);
+    BlevBufferSizesOutItT d_blev_buffer_sizes =
+      reinterpret_cast<BlevBufferSizesOutItT>(allocations[MEM_BLEV_BUFFER_SIZES]);
+    BlevBufferTileOffsetsOutItT d_blev_block_offsets =
+      reinterpret_cast<BlevBufferTileOffsetsOutItT>(allocations[MEM_BLEV_BUFFER_TBLOCK]);
+
+    // Kernels' grid sizes
+    uint32_t init_grid_size         = DivideAndRoundUp(num_tiles, INIT_KERNEL_THREADS);
+    uint32_t scan_grid_size         = DivideAndRoundUp(num_tiles, INIT_KERNEL_THREADS);
```

Comment:
Should these variables use one of the `OffsetT` types? Currently, types are implicitly converted throughout the dispatch layer.

Reply:
Good point 👍  I've unified this to use `BlockOffsetT` now.